auth/src/main/java/org/aerogear/mobile/security/AsyncSecurityCheckExecutor.java

```diff
@@ -91,26 +91,41 @@ public AsyncSecurityCheckExecutor build() {
      * Executes the checks asynchronously.
      *
      * Returns a {@link Map} containing the results of each executed test (a {@link Future}).
-     * The key of the map will be the output of {@link SecurityCheck#getName()}, while the value will be
+     * The key of the map will be the output of {@link SecurityCheck#getType()}, while the value will be
      * a {@link Map} of {@link Future} with the {@link SecurityCheckResult} of the check.
      *
      * @return {@link Map}
      */
     public Map<String, Future<SecurityCheckResult>> execute() {
-
-        final MetricsService metricsService = getMetricsService();
         final Map<String, Future<SecurityCheckResult>> res = new HashMap<>();
 
-        for (final SecurityCheck check : getChecks()) {
-            res.put(check.getName(), (executorService.submit(() -> {
-                final SecurityCheckResult result =  check.test(getContext());
-                if (metricsService != null) {
-                    metricsService.publish(new SecurityCheckResultMetric(result));
-                }
+        final Collection<SecurityCheck> checks = getChecks();
+
+        for (final SecurityCheck check : checks) {
+            final Future<SecurityCheckResult> future = executorService.submit(() -> {
+                SecurityCheckResult result = check.test(getContext());
+                tallyResultsAndPublish(result);
                 return result;
-            })));
+            });
+            res.put(check.getType(), future);
         }
 
         return res;
     }
+
+    private synchronized void tallyResultsAndPublish(SecurityCheckResult result) {
+        MetricsService metricsService = this.getMetricsService();
+        if (metricsService == null) {
+            return;
+        }
+
+        int checkCount = getChecks().size();
+        if (results == null) {
+            results = new ArrayList<>(checkCount);
```

Comment:
This way if the called will call 2 times the execute method, metrics will be messed up and some metric will be lost.
With the proposed approach, we would not have this issue.

Reply:
I thought the method being `synchronized` would prevent this.

Reply:
Nope. The result array is a member variable. If I call the execute two times, I will add the results twice in the same array

Reply:
Ah, I see what you mean, the executor interface doesn't prevent it being executed multiple times. :+1: